logstash/lib/logstash/outputs/loki.rb

```diff
@@ -0,0 +1,262 @@
+  end
+
+  def validate_ssl_key
+    if !@key.is_a?(OpenSSL::PKey::RSA) && !@key.is_a?(OpenSSL::PKey::DSA)
+      raise "Unsupported private key type #{key.class}"
+    end
+  end
+
+  def ssl_opts(uri)
+    opts = {
+      use_ssl: uri.scheme == 'https'
+    }
+
+    if !@cert.nil? && !@key.nil?
+      opts = opts.merge(
+        verify_mode: OpenSSL::SSL::VERIFY_PEER,
+        cert: @cert,
+        key: @key
+      )
+    end
+
+    unless @ca_cert.nil?
+      opts = opts.merge(
+        ca_file: @ca_cert
+      )
+    end
+    opts
+  end 
+
+  def run()
+	  min_wait_checkfrequency = 1/1000 #1 millisecond
+	  max_wait_checkfrequency = @batch_wait / 10
+	  if max_wait_checkfrequency < min_wait_checkfrequency
+		  max_wait_checkfrequency = min_wait_checkfrequency
+    end
+
+    max_wait_check = Concurrent::Channel.tick(max_wait_checkfrequency)
+    loop do
+      Concurrent::Channel.select do |s|
+        s.take(@entries) { |e|
+          if !@batches.has_key?(e.tenant_id)
+            @batches[e.tenant_id] = Batch.new(e)
+            next
+          end
+
+          batch = @batches[e.tenant_id]
+          line = e.entry['line']
+          if batch.size_bytes_after(line) > @batch_size
+            send(e.tenant_id, batch)
+            @batches[e.tenant_id] = Batch.new(e)
+            next
+          end
+          batch.add(e)
+        }
+        s.take(max_wait_check) {
+          # Send all batches whose max wait time has been reached
+          @batches.each { |tenant_id, batch|
+            if batch.age() < @batch_wait
+              next
+            end
+            send(tenant_id, batch)
+            @batches.delete(tenant_id)
+          }
+        }
+      end
+    end 
+  end
+
+  public
+  def receive(event)
+    labels = {}
+    event_hash = event.to_hash
+    lbls = extract_labels(event_hash, labels, "")
+    data_labels = lbls.merge(@external_labels)
+
+    entry_hash = {
+      "ts" => event.get("@timestamp").to_i * (10**9) +  Time.new.nsec,
+      "line" => event.get("message").to_s
+    }
+
+    @Channel.go do
+      @entries << Entry.new(@tenant_id, data_labels, entry_hash)
+    end 
+  end
+
+  def format_key(key)
+    if key[0] == "@"
+      key = key[1..-1]
+    end
+    return key
+  end 
+
+  def extract_labels(event_hash, labels, parent_key)
+    event_hash.each{ |key,value|
+      if !@exclude_labels.include?(key) && !value.is_a?(Array)
+        key = format_key(key)
+        if value.is_a?(Hash)
+          if parent_key != ""
+            extract_labels(value, labels, parent_key + "_" + key)
+          else  
+            extract_labels(value, labels, key)
+          end
+        else
+          if parent_key != ""
+            labels[parent_key + "_" + key] = value.to_s
+          else 
+            labels[key] = value.to_s
+          end
+        end
+      end
+    }
+    return labels
+  end
+
+  def close
+    @logger.info("Closing loki-logstash-output")
+    @entries.close 
```

Comment:
Should you ensure forming batches are flush before closing ? I this should be tested too.

Reply:
Yeah, missed that. Will include flushing

Reply:
You need to also find a way stop the goroutine run created at register time.

Reply:
Logstash is shutingdown correctly. Let me re check still

Reply:
That goroutine create in the register should be stopped. Since you already closing entries you need to read that signal in the goroutine.

```
 Concurrent::Channel.select do |s|
        s.take(@entries) { |e|
        if !e {
           return // exit the goroutine.
        }
       }
```

seee https://github.com/ruby-concurrency/concurrent-ruby/blob/master/examples/go-by-example-channels/closing-channels.rb

Reply:
Will make the change. Thanks

Reply:
I actually tested it. `s.take(@entries)` would not have any value when @entries is closed. So, it would not actually go inside that part of the code.

Reply:
closing the `entries` and `max_wait_check` channels will make the go channel not do anything. I see that when there is nothing to receive in the `select` the loop is automatically stopped. Even the examples they have shown indicates the same thing. We don't have to stop the thread explicitly. However, when we list the active threads it still shows the `Channel.go` thread is active. Let me check how we can stop this thread. Not got anything still

Reply:
Verified with Concurrent/Ruby gitter channel. Got below reply 

```
Concurrent ruby typically runs things on a global thread pool that will manage, reuse and recycle threads on its own. So if you're confident that the thread isn't doing any work, it shouldn't need to be actively killed.
```

So, I think closing the channels is enough.